### DISCUSSION TITLE
What method do you use for obtaining remote data at WebAssembly project?

### DISCUSSION THREAD

**Discussion by I4-PJ** [AvaloniaUI/Avalonia#13315] [Q&A] [answered]
Hello,

I am just a newbie to Avalonia and its WebAssembly. I would like to ask if anyone is already using it together with using RestAPI or direct access to database (MS SQL)? I fact, at our internally developed server services are 3 options on how to get data remotely from the server:
1. Using proprietary TCP/IP communication protocol (have its benefits). The client initiate TCP stream connection to the server.
2. Using RestAPI (implemented via HttpClient / HttpListener at server side)
3. Direct data access to MS SQL database.

As of now, I am not lucky at all having any method working when targeting WebAssembly. However, all mentioned methods work with Desktop being a target. 

Is there anyone having similar issues or what method for acquiring remote data do you use?

Thanks a lot for hints,
PJ.

**Comment by timunie** [comment]
I believe this is an issue for .NET in general, not Avalonia. So probably you can query SO or another forum for it. 

For example: 
- https://stackoverflow.com/questions/63996294/how-to-connect-blazor-webassembly-to-database

**Comment by maxkatz6** [comment]
> is already using it together with using RestAPI

Use HttpClient or any wrapper on top of it. 

> access to database (MS SQL)

No, not possible. Ever.

> Using proprietary TCP/IP communication protocol (have its benefits).

Not possible ever either.

**Reply by maxkatz6** [reply, marked as answer]
Just to clarify, it has nothing to do with WASM, Avalonia, or even .NET.
The browser itself doesn't allow access to TCP connections at all. There is no API for this.
The same as with JS on the web, you have only one choice of using HTTP-based protocols (HTTP1/2/3, WebSockets, HTTP Streaming...). Any non-HTTP protocol has to be emulated on top of HTTP to work in the browser as well, like gRPC (google: gRPC over HTTP2).

Not to mention, in some cases, direct access to the database from the web app would be a disaster for security.

Also, as it could be the next potential problem, be aware of CORS-related issues. Browser doesn't allow you to make HTTP requests to just any remote server. The remote server needs to approve your client's origin address.

**Reply by maxkatz6** [reply]
> HttpListener at server side

While it will work for sure, I would recommend using more feature-complete frameworks like asp.net core. It also will handle some nuances like CORS for you (just need to configure). With minimal API support, asp.net core can be quite easy to write and maintain, if you don't need a lot of endpoints.

**Reply by I4-PJ** [reply]
Thanks a lot Max for a detailed clarification. Understood TCP and direct access not being an option. From my perspective, RAPI is perfectly fine we have just to handle some issues. As you mentioned using Httpclient, unfortunatelly, we can not have even a connection initiated for some reason. But thats another story. From web browser just typing remote RAPI address everything works perfectly fine.

We are avoiding ASP etc. as our service must be as simple in management as possible with keeping all the functionalities. Our implementations are not at public networks but at industrial ones after several firewall layers. 

Thanks a lot,
PJ.
